src/index.js

```
@@ -150,12 +150,18 @@ class IOpipeWrapperClass {
   invoke() {
     this.runHook('pre:invoke');
     try {
-      return this.userFunc.call(
+      const result = this.userFunc.call(
         this.originalIdentity,
         this.event,
         this.context,
         this.callback
       );
+      if (result && result.then && result.catch) {
+        return new Promise(() => this.callback(null, () => result))
+          .then(value => value)
+          .catch(err => this.callback(err));
+      }
+      return result;
     } catch (err) {
       this.sendReport(err, () => this.originalCallback(err));
       return err;
```

Shouldn't this rethrow?

---

When I trigger an error there, relying on the callback looks to be the better option--the stack trace is preserved and delivered to the dashboard.  Re-throwing, on the other hand, causes tests and build to fail.

---

But what gets passed back to AWS? Don't think we want to catch uncaught errors, as that will prevent AWS from detecting a retry scenario, for example. iopipe-python re-raises uncaught exceptions.

---

you can see what we've done for callback based results in the block that follows this one:

```
    } catch (err) {
      this.sendReport(err, () => this.originalCallback(err));
      return err;
```

This makes me wonder if the right behavior is to return the error rather than `return result` when we catch an error?

It doesn't seem we're re-throwing?

---

@kolanos If I trigger errors (in this case, with `with context.fail('ouch')`), I see a full stack going up to AWS: https://dashboard.iopipe.com/project/97ad3ff5-506f-4072-a2a8-06ac9a7c901c/functions/arn:aws:lambda:local:0:function:iopipe-barstool-test-1-dev-minimalTest/invocation/AWuuQdw5E5SfY795__gt#timeframe=4h

---

@mrickard We need to make sure AWS Lambda knows this is an error so it can retry if needed. I know that rethrowing will break tests, but that's because the tests may be testing incorrect behavior.

@ewindisch It appears we're just returning the error object, not rethrowing and letting AWS Lambda handle the exception the way that iopipe-python does.

---

@ewindisch If the result being handled is a promise, the handler won't hit that `return result;`.  Just above the quoted block, what's being returned is the promise that includes the catch block. Currently that isn't returning the error, but I'll update it so that it does.

---

Currently sending the report and throwing from the promise based on other conversations in this thread.